### PR TITLE
[BitcoinCash branch] cashaddr

### PR DESCRIPTION
**Pull request by ptschip**

### CODE REVIEW

**Comment by ptschip** [comment]
@gandrewstone @sickpig cashaddr is ready for testing

**Comment by ptschip** [comment]
bitcoin-QT is crashing on startup, bitcoind is fine though.  Seems to be happening after
665e38e was applied.

**Comment by ptschip** [comment]
Fixed the crash on starup...testing around a bit and finding the coin freeze doesn't seem to be working when making a coin freeze address...i'll have to look into that next week.

File: depends/README.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ For example:
 
 A prefix will be generated that's suitable for plugging into Bitcoin's
 configure. In the above example, a dir named x86_64-w64-mingw32 will be
-created. To use it for Bitcoin:
+created. To use it for bitcoin-abc:
```
Comment (gandrewstone) [inline depends/README.md]: remove this change

File: doc/release-notes.md [hunk 2]
```diff
@@ -5,6 +5,7 @@ Bitcoin Unlimited version 0.12.1 is now available from:
 This is a new minor version release, including ........,
 various bugfixes and updated translations.
 
+<<<<<<< HEAD
```
Comment (gandrewstone) [inline doc/release-notes.md:8]: looks like merge annotation accidentally got in

File: doc/release-notes.md [hunk 3]
```diff
@@ -122,3 +123,10 @@ Thanks to everyone who directly contributed to this release:
 
 
 As well as everyone that helped translating on [Transifex](https://www.transifex.com/projects/p/bitcoin/).
+=======
+ - Remove the newdaaactivationtime configuration.
+ - Do not use the NODE_BITCOIN_CASH service bit for preferencial peering anymore.
+ - Only connect to node using the cash magic.
+ - Remove indicator mentionning if a node uses the cash magic getpeerinfo RPC.
+ - Add support for the new cashaddr format. The `-usecashaddr` flag can be used to select which format is used when presenting addresses to users. By default, Bitcoin ABC will keep using the old format until Jan, 14 and then switch to the new format. Both format are now accepted as input.
+>>>>>>> 7557637... Activate cashaddr on Jan, 14
```
Comment (gandrewstone) [inline doc/release-notes.md]: merge annotation and change Bitcoin ABC to "this client".

File: src/bitcoin-tx.cpp [hunk 4]
```diff
@@ -298,19 +308,30 @@ static void MutateTxAddOutData(CMutableTransaction &tx, const string &strInput)
...
 
     // extract and validate script
-    string strScript = strInput.substr(pos + 1, string::npos);
-    CScript scriptPubKey = ParseScript(strScript); // throws on err
+    std::string strScript = vStrInputParts[1];
+    CScript scriptPubKey = ParseScript(strScript);
+
+    // Extract FLAGS
+    bool bScriptHash = false;
+    if (vStrInputParts.size() == 3)
```
Comment (gandrewstone) [inline src/bitcoin-tx.cpp]: If the string has 4 parts this code will ignore the flags.  To maintain compatibility with the previous behavior, we would ignore additional sections.  so this should be: if (vStrInputParts.size() > 2)

File: src/config.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+// Dummy for subclassing in unittests
+class DummyConfig : public Config {
+public:
+    const CChainParams &GetChainParams() const override;
+    void SetCashAddrEncoding(bool) override {}
+    bool UseCashAddrEncoding() const override { return false; }
+};
+
+// Temporary woraround.
```
Comment (gandrewstone) [inline src/config.h:43]: why is this a workaround, what should really be done?

File: depends/README.md (hunk 1 above)
Comment (ptschip) [reply depends/README.md]: done

File: doc/release-notes.md (hunk 2 above)
Comment (ptschip) [reply doc/release-notes.md:8]: fixed

File: doc/release-notes.md (hunk 3 above)
Comment (ptschip) [reply doc/release-notes.md]: fixed

File: src/bitcoin-tx.cpp (hunk 4 above)
Comment (ptschip) [reply src/bitcoin-tx.cpp]: fixed

File: src/config.h (hunk 5 above)
Comment (ptschip) [reply src/config.h:43]: I didn't put the text there so I don't know why it was considered a workaround and I don't see how it could be different...

Comment (dagurval) [reply src/config.h:43]: Not my comment either, but I believe the intention is not have a global getter, but rather pass the global config instance as argument to everywhere it's needed.

File: contrib/rpm/bitcoin.spec [hunk 6]
```diff
@@ -0,0 +1,440 @@
+%define bdbv 4.8.30
```
Comment (sickpig) [inline contrib/rpm/bitcoin.spec]: rpm spec file is not a bad add at all but I think we should mv it on its own PR

File: doc/release-notes.md [hunk 7]
```diff
@@ -36,79 +37,27 @@ earlier.
...
-- RPC `decodescript`
-- REST `/rest/tx/` (JSON format)
-- REST `/rest/block/` (JSON format when including extended tx details)
-- `bitcoin-tx -json`
-
 ### Configuration and command-line options
 
 ### Block and transaction handling
 
+Remove the newdaaactivationtime configuration.
```
Comment (sickpig) [inline doc/release-notes.md]: This change is not related to cashaddr

File: doc/release-notes.md [hunk 8]
```diff
@@ -36,79 +37,27 @@ earlier.
...
+
 ### P2P protocol and network code
 
-The p2p alert system has been removed in #7692 and the 'alert' message is no longer supported.
+Only connect to node using the cash magic.
```
Comment (sickpig) [inline doc/release-notes.md]: ditto

File: src/Makefile.am [hunk 9]
```diff
@@ -95,13 +97,15 @@ BITCOIN_CORE_H = \
   compat/endian.h \
   compat/sanity.h \
   compressor.h \
+  config.h \
```
Comment (sickpig) [inline src/Makefile.am:100]: why we need this one?

File: src/Makefile.am [hunk 10]
```diff
@@ -477,7 +485,6 @@ clean-local:
 	-$(MAKE) -C secp256k1 clean
 	-$(MAKE) -C univalue clean
 	-rm -f leveldb/*/*.gcda leveldb/*/*.gcno leveldb/helpers/memenv/*.gcda leveldb/helpers/memenv/*.gcno
-	-rm -f config.h
```
Comment (sickpig) [inline src/Makefile.am:480]: seems wrong to me

File: src/allowed_args.cpp [hunk 11]
```diff
@@ -439,7 +439,9 @@ static void addWalletOptions(AllowedArgs &allowedArgs)
             _("Delete all wallet transactions and only recover those parts of the blockchain through -rescan on "
               "startup") +
                 " " +
-                _("(1 = keep tx meta data e.g. account owner and payment request information, 2 = drop tx meta data)"));
+                _("(1 = keep tx meta data e.g. account owner and payment request information, 2 = drop tx meta data)"))
+        .addArg("usecashaddr", optionalBool,
+            _("Use Bitcoin Cash Address for destination encoding (Activates by default Jan 14, 2017)"));
```
Comment (sickpig) [inline src/allowed_args.cpp:444]: this means we need to release in a few days right?

File: src/cashaddr.cpp [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+
+    return ret;
+}
+
+/**
+ * Decode a cashaddr string.
+ */
+std::pair<std::string, data> Decode(const std::string &str,
+                                    const std::string &default_prefix) {
+    // Go over the string and do some sanity checks.
```
Comment (gandrewstone) [inline src/cashaddr.cpp]: what's the longest cashaddr?  Let's add this at the top:
if (str.size() > X) return {};

That way the rest of the routine is protected from someone sending a very long string...

File: src/config.h (hunk 5 above)
Comment (gandrewstone) [reply src/config.h:43]: I'm ok with using a global since I can't conceive of a good reason for running multiple configs simultaneously.  Let's remove the comment.

File: src/qt/locale/bitcoin_be_BY.ts [hunk 13]
```diff
@@ -410,8 +414,8 @@
         <translation>Паказаць спіс адрасоў і метак для прымання</translation>
     </message>
     <message>
-        <source>Open a %1: URI or payment request</source>
-        <translation>Адкрыць %1: URI ці запыт плацяжу</translation>
+        <source>Open a %1 URI or payment request</source>
+        <translation>Адкрыць біткойн: URI ці запыт плацяжу</translation>
```
Comment (gandrewstone) [inline src/qt/locale/bitcoin_be_BY.ts]: the %1 got lost...

File: src/qt/locale/bitcoin_ko_KR.ts [hunk 14]
```diff
@@ -402,8 +402,8 @@
         <translation>비트코인 코어</translation>
     </message>
     <message>
-        <source>Request payments (generates QR codes and %1: URIs)</source>
-        <translation>지불 요청하기 (QR코드와 비트코인이 %1: URIs)</translation>
+        <source>Request payments (generates QR codes and %1 URIs)</source>
+        <translation>지불 요청하기 (QR코드와 비트코인이 생성됩니다: URIs)</translation>
```
Comment (gandrewstone) [inline src/qt/locale/bitcoin_ko_KR.ts]: %1

File: src/script/ismine.h [hunk 15]
```diff
@@ -19,7 +19,8 @@ class CScript;
 enum isminetype
 {
     ISMINE_NO = 0,
-    //! Indicates that we don't know how to create a scriptSig that would solve this if we were given the appropriate private keys
+
+
```
Comment (gandrewstone) [inline src/script/ismine.h]: probably good to keep this comment...

Comment (ptschip) [reply src/script/ismine.h]: fixed

File: src/qt/locale/bitcoin_ko_KR.ts (hunk 14 above)
Comment (ptschip) [reply src/qt/locale/bitcoin_ko_KR.ts]: fixed

File: src/qt/locale/bitcoin_be_BY.ts (hunk 13 above)
Comment (ptschip) [reply src/qt/locale/bitcoin_be_BY.ts]: fixed

File: src/cashaddr.cpp (hunk 12 above)
Comment (ptschip) [reply src/cashaddr.cpp]: I tried just using a length of 42 + prefixlength + 1 for the separator...it generally works but many of tests then fail.  Will have to research this a bit.

File: src/chainparams.cpp [hunk 16]
```diff
@@ -265,6 +266,7 @@ class CUnlParams : public CChainParams
             boost::assign::list_of(0x42)(0x69)(0x67)(0x20).convert_to_container<std::vector<unsigned char> >();
```
Comment (sickpig) [inline src/chainparams.cpp:266]: why there's a different way to set up `EXT_PUBLIC_KEY` and `EXT_SECRET_KEY` in respect to what we do for bch testnet a few lines below (L368)?

File: src/config.cpp [hunk 17]
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2017 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
```
Comment (sickpig) [inline src/config.cpp:4]: wonder if we could put the content of this file in `src/globals.cpp` in a future PR?

File: src/qt/guiconstants.h [hunk 18]
```diff
@@ -56,7 +56,7 @@ static const int TOOLTIP_WRAP_THRESHOLD = 80;
 static const int MAX_URI_LENGTH = 255;
 
 /* QRCodeDialog -- size of exported QR Code image */
-#define EXPORT_IMAGE_SIZE 256
+#define QR_IMAGE_SIZE 350
```
Comment (sickpig) [inline src/qt/guiconstants.h:59]: why this change in image size?

Comment (ptschip) [reply src/qt/guiconstants.h:59]: it was a carry over from the ABC PR...i think it's fine and actually a bigger image is i think better.  I tried scanning the image from my computer with my cell and it works fine whereas in the past I sometimes had issues (if my memory serves)

File: src/config.cpp (hunk 17 above)
Comment (ptschip) [reply src/config.cpp:4]: Good question...I think it's intent is correct and helps to pass the correct config based on what environment we're running in rather than a global var which we need regardless of environment

File: src/chainparams.cpp (hunk 16 above)
Comment (ptschip) [reply src/chainparams.cpp:266]: It's an older definition..the ones carried over from ABC were newer and look better , but there's nothing wrong with what is there.  Maybe in a cleanup PR we can update the old definitions so everything is uniform.

File: src/allowed_args.cpp (hunk 11 above)
Comment (ptschip) [reply src/allowed_args.cpp:444]: yes, ABC is activating by default on the 14th.  I think we need to get this out there as well ASAP.

File: src/Makefile.am (hunk 10 above)
Comment (ptschip) [reply src/Makefile.am:480]: config.h was being deleted when you ran "make clean"..this was incorrect and should not be happening.

File: src/Makefile.am (hunk 9 above)
Comment (ptschip) [reply src/Makefile.am:100]: this is the new config that we had to carry over from ABC to make cashaddr work.

File: doc/release-notes.md (hunk 8 above)
Comment (ptschip) [reply doc/release-notes.md]: fixed.

File: doc/release-notes.md (hunk 7 above)
Comment (ptschip) [reply doc/release-notes.md]: fixed

File: contrib/rpm/bitcoin.spec (hunk 6 above)
Comment (ptschip) [reply contrib/rpm/bitcoin.spec]: fixed

**Comment by ptschip** [comment]
@sickpig ok, took out the bitcoin.spec.

File: src/qt/guiconstants.h (hunk 18 above)
Comment (dagurval) [reply src/qt/guiconstants.h:59]: In ABC, the text representation of the address is embedded in the QR code. The image needed to be increased to fit the text.

**Comment by gandrewstone** [comment]
@ptschip, in the send dialog, the Public Label field has an & in front of it.  This is supposed to indicate a hotkey but somehow it is being displayed instead.  We can fix this in a subsequent PR if nothing else pops up...
